docs/api/web-contents.md

```diff
@@ -265,6 +265,30 @@ this purpose.
 
 Calling `event.preventDefault()` will prevent the navigation.
 
+#### Event: 'will-frame-navigate'
+
+Returns:
+
+* `event` Event
+* `url` String
+* `isMainFrame` Boolean
+* `frameProcessId` Integer
+* `frameRoutingId` Integer
```

Comment:
The design of the `will-navigate` API was arguably an extensibility mistake. I wonder whether we should be repeating the same here. Our API design best practices goes over this: https://github.com/electron/governance/blob/main/wg-api/best-practices.md#how-will-this-api-be-extended-in-the-future

I'd like to start a discussion as to whether we should break from consistency and instead make this an extensible design.
```ts
interface WebContents {
  on(eventName: 'will-frame-navigate', (e: Event, details: WillNavigateDetails) => void); 
}

interface WillNavigateDetails {
  url: string;
  frame: WebFrameMain;
};
```

Using an instance of `WebFrameMain` would provide each of the frame-specific details.

Reply:
The downside of this approach is that `WebFrameMain` would not be available in the renderer for the `<webview>`'s equivalent event. I thought we had a solution for this with another event which @miniak fixed, do you recall what that was?

Reply:
I like the idea of keeping the URL as a positional parameter for consistency. What if we kept the `WillFrameNavigateDetails` object similar to the parameters of other navigation events, just as an object rather than positional args? That way it could still have the same shape for `<webview>`, and it's easy to enough to use `webFrameMain.fromId` to retrieve the actual `WebFrameMain` object.

```typescript
interface WebContents {
  on(eventName: 'will-frame-navigate', (e: Event, url: string, details: WillFrameNavigateDetails) => void); 
}

interface WillFrameNavigateDetails {
  isMainFrame: boolean;
  frameProcessId: number;
  frameRoutingId: number;
}
```

Reply:
@samuelmaddock any thoughts on keeping `frameProcessId` and `frameRoutingId` rather than directly exposing the `WebFrameMain`?

Reply:
Would it be possible to transform the details before emitting on the `<webview>`? I'm fine with keeping the originally proposed API on the webview since WebFrameMain isn't available there anyway.

Reply:
I'll take a look. It was easy to just forward parameters by adding an entry in `web-view-events.ts` but I can dig around under the hood and see if I can do a more custom transformation instead